static/js/controllers/contacts-content.js

```diff
@@ -114,6 +121,12 @@ angular.module('inboxControllers').controller('ContactsContentCtrl',
 
     // exposed solely for testing purposes
     this.setupPromise = $q.resolve().then(function() {
+      if ($stateParams.noDebounce) {
+        debouncedReloadContact = reloadContact;
+      } else if ($stateParams.debounceWait) {
+        debouncedReloadContact = _.debounce(reloadContact, $stateParams.debounceWait);
+      }
+
```

Comment:
As far as I can tell this new section is only used in testing - is that right?

I'm wary of modifying production code to support test code unless we absolutely have to. The problem is you're overwriting the production debouncing code so you're no longer testing the logic. You've mitigated the issue by testing a small `debounceWait` value which is good but it's still something I'd like to avoid if possible.

When unit testing becomes too hard it usually means your code is too complicated and you should pull some of it out as services that you can mock in your test. I think it would really help here to pull out a `Debounce` service (or whatever name you prefer). You could then mock that service to invoke the callback whenever you like, either immediately (equivalent to `$stateParams.noDebounce`) or after a short wait (as in `$stateParams.debounceWait`).

This `Debounce` could be a simple wrapper for the `_.debounce` function but I think there are more improvements we make if we use the `$timeout` service directly. Firstly the `$timeout` service requests that angular digest and update the UI which is useful here. Secondly `$timeout` has a `cancel` function which can be used when destroying this controller so we don't need the `hasChanges` variable any more.

I'm thinking something [like this](https://github.com/shahata/angular-debounce/blob/master/angular-debounce.js#L4-L33). I'll leave it up to you whether to include the `angular-debounce` dependency or write our own.

Reply:
Yes, you're correct. 
I have created a Debounce Service which uses Angular's $timeout and supports the same functionality, with the added logic of having a maximum delay after which the callback will be called. 
Please re-review.